File: FlattenBinaryTreetoLinkedList2.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def flatten(self, root):
        """
        :type root: TreeNode
        :rtype: void Do not return anything, modify root in-place instead.
        """
        if not root:
            return
        self.flatten(root.left)
        self.flatten(root.right)
        left = root.left
        right = root.right
        
        if not left:
            return
        
        root.right = left
        root.left = None
        while left.right:
            left = left.right
        left.right = right
        
if __name__=='__main__':
    root=TreeNode(1);
    a=TreeNode(2);
    b=TreeNode(3);
    c=TreeNode(4);
    d=TreeNode(5);
    e=TreeNode(6);
    
    root.left=a;
    root.right=d;
    a.left=b;
    a.right=c;
    d.right=e;
    
    solution = Solution()
    solution.flatten(root)
    
    while root:
        print(root.val)
        root = root.right

File: FlattenBinaryTreetoLinkedList2.cpp
#include<iostream>
#include<vector>

using namespace std;

struct TreeNode{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    TreeNode *pre=NULL;
    void flatten(TreeNode* root) {
        
        if(root==NULL) return;
        
        flatten(root->right);
        flatten(root->left);
        
        root->right=pre;
        root->left=NULL;
        
        pre=root;
        
    }
};

int main(){
    TreeNode *root=new TreeNode(1);
    TreeNode *a=new TreeNode(2);
    TreeNode *b=new TreeNode(3);
    TreeNode *c=new TreeNode(4);
    TreeNode *d=new TreeNode(5);
    TreeNode *e=new TreeNode(6);
    TreeNode *f=new TreeNode(7);
    
    root->left=a;
    root->right=b;
    a->left=c;
    c->left=d;
    b->left=e;
    b->right=f;
    
    //root->left=a;
    //root->right=d;
    //a->left=b;
    //a->right=c;
    //d->right=e;
    
    Solution *solution=new Solution();
    solution->flatten(root);
    
    while(root!=NULL){
        cout<<root->val<<" ";
        root=root->right;
    }
    
    return 0;
}
